### ISSUE TITLE
False negative in use_self pedantic lint

### ISSUE THREAD

**Issue by flip1995** [rust-lang-nursery/rust-clippy#3231]
cc https://github.com/rust-lang-nursery/rust-clippy/pull/3224#discussion_r220516300

The `use_self` lint does not trigger on existential types with generic type specification:

```rust
struct Foo;

impl Foo {
    fn f(foos: &[Foo]) -> impl Iterator<Item=&Foo> {
    //           ^^^ triggers here            ^^^ should also trigger here
        foos.iter()
    }
}
```

[Playground](https://play.rust-lang.org/?gist=12a72329dae63017b75cfc46f0872737&version=nightly&mode=debug&edition=2015)
